ee/hogai/graph/insights/nodes.py

```diff
@@ -0,0 +1,481 @@
+
+        Side effects:
+            Automatically removes expired cache entries to prevent memory bloat (keep last 100 entries)
+        """
+        if cache_key in self._semantic_cache:
+            cached_data, timestamp = self._semantic_cache[cache_key]
+            if time.time() - timestamp < self._cache_ttl:
+                return cached_data, True
+            else:
+                del self._semantic_cache[cache_key]
+        return None, False
+
+    def _cache_semantic_result(self, cache_key: str, result: BestInsightSelection) -> None:
+        """Cache semantic filtering result with timestamp."""
+        # Simple cache management - keep last 100 entries
+        if len(self._semantic_cache) > 100:
+            # Remove oldest entries (basic LRU)
+            oldest_key = min(self._semantic_cache.keys(), key=lambda k: self._semantic_cache[k][1])
+            del self._semantic_cache[oldest_key]
+
+        self._semantic_cache[cache_key] = (result, time.time())
+
+    def run(self, state: AssistantState, config: RunnableConfig) -> PartialAssistantState | None:
+        start_time = time.time()
+        search_query = state.search_insights_query
+        conversation_id = config["configurable"]["thread_id"]
+
+        cache_hits = 0
+        cache_misses = 0
+
+        try:
+            results, cache_stats = self._search_insights(search_insights_query=search_query)
+            cache_hits = cache_stats.get("hits", 0)
+            cache_misses = cache_stats.get("misses", 0)
+
+            # Format the results for display
+            formatted_content = self._format_insight_results(results, search_query)
+
+            # Log performance metrics
+            execution_time = time.time() - start_time
+            total_cache_attempts = cache_hits + cache_misses
+            cache_hit_rate = round(cache_hits / max(total_cache_attempts, 1), 2) if total_cache_attempts > 0 else 0
+
+            self.logger.info(
+                f"Insight search completed",
+                extra={
+                    "team_id": self._team.id,
+                    "conversation_id": conversation_id,
+                    "query_length": len(search_query) if search_query else 0,
+                    "results_count": len(results),
+                    "execution_time_ms": round(execution_time * 1000, 2),
+                    "used_semantic_filtering": bool(search_query),
+                    "cache_hit_rate": cache_hit_rate,
+                    "cache_hits": cache_hits,
+                    "cache_misses": cache_misses,
+                },
+            )
+
+            return PartialAssistantState(
+                messages=[
+                    AssistantToolCallMessage(
+                        content=formatted_content,
+                        tool_call_id=state.root_tool_call_id,
+                        id=str(uuid4()),
+                    ),
+                ],
+                # Reset state values
+                search_insights_query="",
```

Comment:
You need to reset `root_tool_call_id` 
```suggestion
                search_insights_query=None,
                root_tool_call_id=None
```